crates/gpui/src/platform/linux/wayland/clipboard.rs

```diff
@@ -0,0 +1,164 @@
+use std::io::{Read, Write};
+use std::os::fd::{AsRawFd, BorrowedFd, OwnedFd};
+
+use anyhow::{bail, Result};
+use filedescriptor::{FileDescriptor, Pipe};
+use wayland_backend::client::ObjectId;
+use wayland_client::protocol::wl_data_offer::WlDataOffer;
+use wayland_client::protocol::wl_data_source::WlDataSource;
+use wayland_client::Proxy;
+
+use super::WaylandClient;
+use crate::ClipboardItem;
+
+pub const TEXT_MIME_TYPE: &str = "text/plain;charset=utf-8";
+
+pub(crate) struct Clipboard {
+    offer: Option<DataOffer>,
+    offer_has_text: bool,
+    offer_is_self: bool,
+    source: Option<WlDataSource>,
+    source_content: Option<String>,
+    self_mime: String,
+}
+
+#[derive(Clone)]
+pub(crate) struct DataOffer {
+    instance: WlDataOffer,
+    mime_types: Vec<String>,
+}
+
+impl DataOffer {
+    pub fn new(offer: WlDataOffer) -> Self {
+        Self {
+            instance: offer,
+            mime_types: vec![],
+        }
+    }
+
+    pub fn id(&self) -> ObjectId {
+        self.instance.id()
+    }
+
+    pub fn add_mime_type(&mut self, mime: String) {
+        self.mime_types.push(mime);
+    }
+}
+
+fn contains_mime_type(offer: &Option<DataOffer>, mime_type: &str) -> bool {
+    match offer {
+        Some(ref offer) => offer.mime_types.iter().any(|m| m == mime_type),
+        _ => false,
+    }
+}
+
+impl Clipboard {
+    pub fn new() -> Self {
+        Self {
+            offer: None,
+            offer_has_text: false,
+            offer_is_self: false,
+            source: None,
+            source_content: None,
+            self_mime: format!("pid/{}", std::process::id()),
```

Comment:
Inspired from glfw, to support self-copy/paste while being blocking, we add this special mime type to the list of supported types. If we see it, we know the offer comes from us and we can skip the IO and just clone the clipboard source content.